Expecta/NSArray+Difference.m

```diff
@@ -0,0 +1,126 @@
+
+  NSMutableArray *mismatches = [NSMutableArray array];
+  NSMutableArray *reorders = [NSMutableArray array];
+  NSMutableSet *usedInsertions = [NSMutableSet set];
+  NSMutableSet *usedRemovals = [NSMutableSet set];
+  for (Change *insertion in diff.insertions) {
+    for (Change *removal in diff.removals) {
+      if (insertion.index == removal.index) {
+        [mismatches addObject:[[ChangePair alloc] initWithInsertion:insertion removal:removal]];
+        [usedInsertions addObject:@(insertion.index)];
+        [usedRemovals addObject:@(removal.index)];
+      }
+      if (insertion.associatedIndex == removal.index) {
+        assert(removal.associatedIndex == insertion.index);
+        [reorders addObject:[[ChangePair alloc] initWithInsertion:insertion removal:removal]];
+        [usedInsertions addObject:@(insertion.index)];
+        [usedRemovals addObject:@(removal.index)];
+      }
+    }
+  }
+
+  NSMutableArray *differences = [NSMutableArray array];
+  for (ChangePair *mismatch in mismatches) {
+    [differences addObject:[NSString stringWithFormat:@"Mismatch at index (%lu)",
+                            mismatch.insertion.index]];
+    if ([mismatch.insertion.object respondsToSelector:@selector(differenceFrom:)]) {
+      NSArray<NSString *> *diffLines = [(id)mismatch.insertion.object differenceFrom:mismatch.removal.object];
+      for (NSString *line in diffLines) {
+        [differences addObject:[NSString stringWithFormat:@"\t%@", line]];
+      }
+    }
+    else {
+      [differences addObjectsFromArray:@[
+        [NSString stringWithFormat:@"\tExpected: (%@)",
+         EXPDescribeObject(mismatch.insertion.object)],
+        [NSString stringWithFormat:@"\tActual: (%@)", EXPDescribeObject(mismatch.removal.object)]
+      ]];
+    }
+  }
+
+  for (ChangePair *reorder in reorders) {
+    [differences addObjectsFromArray:@[
+      [NSString stringWithFormat:@"Wrong index for value: (%@)",
+       EXPDescribeObject(reorder.insertion.object)],
+      [NSString stringWithFormat:@"\tExpected at index (%lu)", reorder.insertion.index],
+      [NSString stringWithFormat:@"\tActual index is (%lu)", reorder.removal.index]
+    ]];
+  }
+
+  for (Change *removal in diff.removals) {
+    if ([usedRemovals containsObject:@(removal.index)]) {
+      continue;
+    }
+
+    [differences addObjectsFromArray:@[
+      @"Actual contains extra item",
+      [NSString stringWithFormat:@"\tindex: (%lu)", removal.index],
+      [NSString stringWithFormat:@"\tvalue: (%@)", EXPDescribeObject(actual[removal.index])]
+    ]];
+  }
+
+  for (Change *insertion in diff.insertions) {
+    if ([usedInsertions containsObject:@(insertion.index)]) {
+      continue;
+    }
+
+    [differences addObjectsFromArray:@[
+      [NSString stringWithFormat:@"Actual misses an item at index (%lu)", insertion.index],
+      [NSString stringWithFormat:@"\tExpected: (%@)", EXPDescribeObject(expected[insertion.index])]
+    ]];
+  }
+
+  return differences.copy;
```

Comment:
why copy? (also this is a method, not a property).

Reply:
to not return a mutable object, but it doesn't seem that important here.